src/gpu/intel/jit/dsl/dsl.hpp

```diff
@@ -0,0 +1,302 @@
+            case kind_t::none:
+                return v2::layout_t(desc, type, 0,
+                        {{col_var, col, 1}, {row_var, row, col}});
+
+            case kind_t::block: {
+                int col_outer = col / col_inner;
+                return v2::layout_t(desc, type, 0,
+                        {{col_var, col_inner, 1}, {row_var, row, col_inner},
+                                {col_var, col_outer, row * col_inner}});
+            }
+
+            case kind_t::vnni: {
+                int row_inner = 4 / t;
+                int row_outer = row / row_inner;
+                int col_outer = col / col_inner;
+                return v2::layout_t(desc, type, 0,
+                        {{row_var, row_inner, 1},
+                                {col_var, col_inner, row_inner},
+                                {row_var, row_outer, col_inner * row_inner},
+                                {col_var, col_outer,
+                                        row_outer * col_inner * row_inner}});
+            }
+
+            // Impossible to hit due to normalization
+            case kind_t::transpose_vnni:
+            default: gpu_assert(false); return {};
+        }
+    }
+
+    // Tile used for 2d Messages
+    tile_t get_2d_tile(type_t type) const {
+        if (transform == kind_t::transpose_vnni) {
+            auto width = pack_size ? pack_size
+                                   : grf_size() / std::max(type.size(), 4);
+            auto height = 32;
+            return {{dims[1], width}, {dims[0], height}};
+        }
+
+        auto width = pack_size ? pack_size : grf_size() / type.size();
+        auto height = 32;
+        return {{dims[0], width}, {dims[1], height}};
+    }
+
+    // Tile used for Block Loads
+    tile_t get_block_tile(type_t type) const {
+        if (transform == kind_t::none) {
+            return {{dims[0], 8 * grf_size() / type.size()}, {dims[1], 1}};
+        } else if (transform == kind_t::block) {
+            return {{dims[0], pack_size ? pack_size : grf_size() / type.size()},
+                    {dims[1], 1}};
+        } else {
+            gpu_assert(false);
+            return {};
+        }
+    }
+
+    static send_cache_hint_t to_ir(ngen::CacheSettingsLSC hint) {
+        switch (hint) {
+            case ngen::CacheSettingsLSC::L1C_L3C:
+                return send_cache_hint_t::load_once;
+            case ngen::CacheSettingsLSC::Default:
+                return send_cache_hint_t::hw_default;
+            default: gpu_assert(false); return send_cache_hint_t::undef;
+        }
+    }
+
+    kind_t transform = kind_t::none;
+    int pack_size = 0;
+    send_cache_hint_t cache_hint = send_cache_hint_t::undef;
+    std::array<pvar_t, 2> dims = {};
+};
+
+struct tensor_t {
+    std::string str() const {
+        std::ostringstream oss;
+        oss << "buffer:    " << buffer.str() << "\n";
+        oss << "layout: " << layout.str() << "\n";
+        return oss.str();
+    }
+
+    expr_t buffer;
```

Comment:
Nit: to align with `ir/core.hpp` naming.
```suggestion
    expr_t buf;
```